File: design-patterns/go/memento/general/originator.go
// originator.go

package main

import "fmt"

type Originator struct {
	state string
}

func NewOriginator() (originator *Originator) {
	originator = &Originator{}
	return
}

func (originator *Originator) GetMementoState(memento *Memento) {
	originator.state = memento.GetState()
}

func (originator *Originator) GetState() (string) {
	return originator.state
}

func (originator *Originator) SetMemento() (*Memento) {
	fmt.Printf("Memento Saved with timestamp => %s\n", originator.state)
	return NewMemento(originator.state)
}

func (originator *Originator) SetState(state string) {
	originator.state = state
}

File: design-patterns/ts/memento/general/originator.ts
// originator.ts

import Memento from "./memento";

class Originator {
    private state: string = '';
    
    public setState(state: string): void {
        this.state = state;
    }

    public getState(): string {
        return this.state;
    }

    public setMemento(): Memento {
        console.log("Memento Saved with timestamp => " + this.state);
        return new Memento(this.state);
    }

    public getMementoState(memento: Memento): void {
        this.state = memento.getState();
    }
}

export default Originator;